rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,606 @@
+          unhealthy_threshold: 2
+
+#### Terraform config example
+
+    resource "teleport_database" "example" {
+      version = "v3"
+      metadata = {
+        name = "example"
+        labels = {
+          "env" = "dev"
+        }
+      }
+      spec = {
+        protocol = "postgres"
+        uri      = "localhost:5432"
+        health_check = {
+          enabled             = true
+          healthy_threshold   = 1
+          unhealthy_threshold = 1
+          interval            = "30s"
+          timeout             = "5s"
+        }
+      }
+    }
+
+#### Discovery config example
+
+Discovery service creates `db` objects and db agents can configure custom health check settings in their dynamic resources matchers.
+
+Dynamic resource matcher config overrides `db.spec.health_check` settings.
+
+    discovery_service:
+      enabled: true
+      discovery_group: "teleport-dev-2"
+      aws:
+        - types:
+            - "rds"
+          regions:
+            - "ca-central-1"
+          tags:
+            "env": "prod"
+
+    db_service:
+      enabled: true
+      resources:
+        - labels:
+            "teleport.dev/origin": "dynamic"
+            "env": "dev"
+          health_check:
+            enabled: false # disable health checks for these databases
+        - labels:
+            "region": "ca-central-1"
+            "teleport.dev/origin": "cloud"
+            "teleport.dev/cloud": "AWS"
+          health_check:
+            enabled: true # enabled with these settings for each DB matched
+            healthy_threshold: 1
+            unhealthy_threshold: 1
+            interval: 30s
+            timeout: 5s
+
+### DB agent behavior
+
+When a DB agent begins proxying a database that has health checks enabled, the agent will asynchronously start a health checker for that database.
+
+When the agent gathers up info for its `db_server` heartbeat, it will get the current health status of the database from its health checker.
+
+The DB agent's heartbeat system currently polls every 5 seconds for local changes to its `db_server` heartbeat.
+
+When any of the `target_health` fields change, the `db_server` heartbeat will be considered different from the old heartbeat, and the new health status will be announced as an updated `db_server`.
+
+When the database is deregistered from the agent, its health checker will be stopped.
+
+When the database is updated, it will be deregistered and then re-registered, effectively restarting its health checker and health status.
+
+### Target address resolution
+
+The health check target address will be taken from the database uri.
+Teleport does not strictly enforce a host:port database uri, so the target host and/or port may need to be resolved.
+The target host and port should generally be resolved in the same way it is for user connections.
+
+As a special case, some databases have multiple endpoints.
+For example, a MongoDB replicaset connection string may look like this:
```

Comment:
This mentioned Mongo - which databases we're scoping this to initially? I don't think I saw this being mentioned unless I missed it.